File: typescript/src/bccae.ts
/**
 * BCCAE V1 — TypeScript client.
 *
 * BCCAEClient wraps the four BCCAE Phase 3 endpoints:
 *   evaluate  → POST /v1/bccae/evaluations   (bccae:evaluate scope)
 *   execute   → POST /v1/bccae/execute        (bccae:execute scope)
 *   revoke    → POST /v1/bccae/revocations    (bccae:revoke scope)
 *   getEvidence → GET /v1/bccae/evidence/:id  (bccae:audit scope)
 *
 * Spec: atlasent-internal/architecture/BCCAE-architecture.md
 * Phase 3 — Execution Assurance. Not a Deploy Gate V1 customer API.
 */

import { AtlaSentError } from "./errors.js";

// ─── Types ────────────────────────────────────────────────────────────────────

export type BccaeActorType = "HUMAN" | "AGENT" | "SERVICE" | "EXTERNAL";
export type BccaeTrustLevel = "L0" | "L1" | "L2" | "L3";
export type BccaeResourceClassification =
  | "PUBLIC"
  | "INTERNAL"
  | "CONFIDENTIAL"
  | "RESTRICTED";
export type BccaeDeploymentEnv = "PROD" | "STAGING" | "DEV" | "TEST";
export type BccaeSecurityPosture = "STANDARD" | "ELEVATED" | "LOCKED";
export type BccaeRequestSource =
  | "AGENT"
  | "API"
  | "INTERNAL"
  | "SCHEDULED"
  | "TRIGGERED";
export type BccaeRevocationTargetType =
  | "PERMIT"
  | "EVALUATION"
  | "ACTOR"
  | "RESOURCE";

export interface BccaeEvaluateInput {
  actor_id: string;
  actor_type: BccaeActorType;
  actor_trust_level: BccaeTrustLevel;
  actor_claims?: Record<string, unknown>;
  action_id: string;
  execution_intent: string;
  /** 64 lowercase hex characters (32 random bytes). */
  caller_nonce: string;
  resource_ref: string;
  resource_type: string;
  resource_classification: BccaeResourceClassification;
  organization_version?: number;
  deployment_env: BccaeDeploymentEnv;
  deployment_region: string;
  security_posture: BccaeSecurityPosture;
  external_signals?: unknown[];
  dependencies?: unknown[];
  policy_version_set?: unknown[];
  request_source?: BccaeRequestSource;
  request_chain_id?: string;
  parent_eval_id?: string;
}

export interface BccaeEvaluateResponse {
  evaluation_id: string;
  envelope_hash: string;
  permit_token: string;
  permit_id: string;
  expires_at: string;
  outcome: "PERMIT" | "PERMIT_WITH_CONDITIONS";
}

export interface BccaeExecuteInput {
  permit_token: string;
  action_id: string;
  resource_ref: string;
}

export interface BccaeExecuteResponse {
  authorized: boolean;
  outcome: "EXECUTION_AUTHORIZED" | "EXECUTION_DENIED";
  permit_id?: string;
  evaluation_id?: string;
  envelope_hash?: string;
  evidence_id?: string | null;
  /** Populated on denial — identifies which gate check failed. */
  check?: string;
  reason?: string;
}

export interface BccaeRevokeInput {
  target_type: BccaeRevocationTargetType;
  target_id: string;
  reason: string;
}

export interface BccaeRevokeResponse {
  revocation_id: string;
  target_type: BccaeRevocationTargetType;
  target_id: string;
  effective_at: string;
}

export interface BccaeEvidenceResponse {
  evidence_id: string;
  org_id: string;
  event_type: string;
  evaluation_id: string | null;
  permit_id: string | null;
  envelope_hash: string | null;
  actor_id: string;
  action_id: string | null;
  resource_ref: string | null;
  outcome: string;
  detail: Record<string, unknown>;
  previous_evidence_id: string | null;
  previous_hash: string | null;
  record_hash: string;
  sequence: number;
  recorded_at: string;
  chain_integrity: {
    hash_intact: boolean;
    expected_hash?: string;
  };
}

export interface BccaeClientOptions {
  /** API key with appropriate bccae:* scopes. */
  apiKey: string;
  /** Override base URL. Defaults to https://api.atlasent.io */
  baseUrl?: string;
  /** Request timeout in ms. Defaults to 10000. */
  timeoutMs?: number;
  /** Inject a custom fetch implementation (testing / edge runtimes). */
  fetch?: typeof globalThis.fetch;
}

// ─── Helpers ──────────────────────────────────────────────────────────────────

const DEFAULT_BASE_URL = "https://api.atlasent.io";
const DEFAULT_TIMEOUT_MS = 10_000;

function enforceTls(url: string): string {
  let parsed: URL;
  try {
    parsed = new URL(url);
  } catch {
    throw new AtlaSentError(
      "BCCAEClient baseUrl is not a valid URL",
      { code: "network" },
    );
  }
  if (parsed.protocol === "http:") {
    // Use parsed.hostname (exact match) not url.includes() to avoid
    // false-positive local detection on query strings like ?x=localhost.
    const h = parsed.hostname;
    if (h !== "localhost" && h !== "127.0.0.1" && h !== "[::1]") {
      throw new AtlaSentError(
        "BCCAEClient baseUrl must use https:// for non-local endpoints",
        { code: "network" },
      );
    }
  }
  return url;
}

/** Generate a cryptographically random 64-char hex nonce (32 bytes). */
export function generateBccaeNonce(): string {
  const bytes = new Uint8Array(32);
  globalThis.crypto.getRandomValues(bytes);
  return Array.from(bytes)
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

// ─── BCCAEClient ──────────────────────────────────────────────────────────────

/**
 * Thin HTTP client for the BCCAE V1 Phase 3 endpoints.
 *
 * Each method maps 1:1 to an edge function:
 *   - {@link BCCAEClient.evaluate}    → v1-bccae-evaluate
 *   - {@link BCCAEClient.execute}     → v1-bccae-execute
 *   - {@link BCCAEClient.revoke}      → v1-bccae-revoke
 *   - {@link BCCAEClient.getEvidence} → v1-bccae-evidence
 *
 * Authorization denials are returned (not thrown). Network errors,
 * invalid API keys, and 5xx responses throw {@link AtlaSentError}.
 *
 * Use {@link generateBccaeNonce} to produce a valid `caller_nonce`.
 */
export class BCCAEClient {
  private readonly apiKey: string;
  private readonly baseUrl: string;
  private readonly timeoutMs: number;
  private readonly fetchImpl: typeof fetch;

  constructor(options: BccaeClientOptions) {
    if (!options.apiKey || typeof options.apiKey !== "string") {
      throw new AtlaSentError("BCCAEClient: apiKey is required", {
        code: "invalid_api_key",
      });
    }
    this.apiKey = options.apiKey;
    this.baseUrl = enforceTls(
      (options.baseUrl ?? DEFAULT_BASE_URL).replace(/\/+$/, ""),
    );
    this.timeoutMs = options.timeoutMs ?? DEFAULT_TIMEOUT_MS;
    this.fetchImpl = options.fetch ?? globalThis.fetch.bind(globalThis);
  }

  async evaluate(input: BccaeEvaluateInput): Promise<BccaeEvaluateResponse> {
    const { body } = await this.post<BccaeEvaluateResponse>(
      "/v1/bccae/evaluations",
      input,
    );
    return body;
  }

  async execute(input: BccaeExecuteInput): Promise<BccaeExecuteResponse> {
    const { body } = await this.post<BccaeExecuteResponse>(
      "/v1/bccae/execute",
      input,
    );
    return body;
  }

  async revoke(input: BccaeRevokeInput): Promise<BccaeRevokeResponse> {
    const { body } = await this.post<BccaeRevokeResponse>(
      "/v1/bccae/revocations",
      input,
    );
    return body;
  }

  async getEvidence(evidenceId: string): Promise<BccaeEvidenceResponse> {
    if (!evidenceId || typeof evidenceId !== "string") {
      throw new AtlaSentError("BCCAEClient: evidenceId is required", {
        code: "bad_request",
      });
    }
    const { body } = await this.get<BccaeEvidenceResponse>(
      `/v1/bccae/evidence/${encodeURIComponent(evidenceId)}`,
    );
    return body;
  }

  // ── HTTP primitives ─────────────────────────────────────────────────────────

  private async post<T>(
    path: string,
    body: unknown,
  ): Promise<{ body: T }> {
    return this.request<T>(path, "POST", body);
  }

  private async get<T>(path: string): Promise<{ body: T }> {
    return this.request<T>(path, "GET", undefined);
  }

  private async request<T>(
    path: string,
    method: "GET" | "POST",
    body: unknown,
  ): Promise<{ body: T }> {
    const url = `${this.baseUrl}${path}`;
    const headers: Record<string, string> = {
      Accept: "application/json",
      Authorization: `Bearer ${this.apiKey}`,
      "User-Agent": "atlasent-bccae-client/1.0",
    };
    if (method === "POST") headers["Content-Type"] = "application/json";

    let response: Response;
    try {
      // codeql[js/server-side-request-forgery] baseUrl validated by enforceTls (https or http+local only).
      response = await this.fetchImpl(url, {
        method,
        headers,
        signal: AbortSignal.timeout(this.timeoutMs),
        ...(method === "POST" ? { body: JSON.stringify(body) } : {}),
      });
    } catch (err) {
      throw new AtlaSentError(
        `BCCAEClient: network error on ${method} ${path}: ${err instanceof Error ? err.message : String(err)}`,
        { code: "network" },
      );
    }

    let responseBody: unknown;
    try {
      responseBody = await response.json();
    } catch {
      throw new AtlaSentError(
        `BCCAEClient: non-JSON response (status ${response.status}) from ${method} ${path}`,
        { code: "network" },
      );
    }

    if (!response.ok) {
      const err = responseBody as Record<string, unknown>;
      const message =
        typeof err?.message === "string"
          ? err.message
          : `BCCAE request failed with status ${response.status}`;
      const code =
        response.status === 401
          ? "invalid_api_key"
          : response.status === 403
            ? "forbidden"
            : response.status === 429
              ? "rate_limited"
              : response.status >= 500
                ? "server_error"
                : "network";
      throw new AtlaSentError(message, { code });
    }

    return { body: responseBody as T };
  }
}

File: python/atlasent/bccae.py
"""BCCAE V1 — Python client.

BCCAEClient wraps the four BCCAE Phase 3 endpoints:

- evaluate    → POST /v1/bccae/evaluations   (bccae:evaluate scope)
- execute     → POST /v1/bccae/execute        (bccae:execute scope)
- revoke      → POST /v1/bccae/revocations    (bccae:revoke scope)
- get_evidence → GET /v1/bccae/evidence/:id   (bccae:audit scope)

Spec: atlasent-internal/architecture/BCCAE-architecture.md
Phase 3 — Execution Assurance. Not a Deploy Gate V1 customer API.
"""

from __future__ import annotations

import secrets
from typing import Any, Literal
from urllib.parse import quote, urlparse, urlunparse

import httpx

from .exceptions import AtlaSentError

DEFAULT_BASE_URL = "https://api.atlasent.io"
DEFAULT_TIMEOUT = 10.0

# ── Types ─────────────────────────────────────────────────────────────────────────────

BccaeActorType = Literal["HUMAN", "AGENT", "SERVICE", "EXTERNAL"]
BccaeTrustLevel = Literal["L0", "L1", "L2", "L3"]
BccaeResourceClassification = Literal[
    "PUBLIC", "INTERNAL", "CONFIDENTIAL", "RESTRICTED"
]
BccaeDeploymentEnv = Literal["PROD", "STAGING", "DEV", "TEST"]
BccaeSecurityPosture = Literal["STANDARD", "ELEVATED", "LOCKED"]
BccaeRequestSource = Literal["AGENT", "API", "INTERNAL", "SCHEDULED", "TRIGGERED"]
BccaeRevocationTargetType = Literal["PERMIT", "EVALUATION", "ACTOR", "RESOURCE"]


def generate_bccae_nonce() -> str:
    """Return a cryptographically random 64-character hex nonce (32 bytes)."""
    return secrets.token_hex(32)


def _enforce_tls(base_url: str) -> str:
    parsed = urlparse(base_url)
    if parsed.scheme == "https":
        pass
    elif parsed.scheme == "http":
        is_local = parsed.hostname in ("localhost", "127.0.0.1", "::1")
        if not is_local:
            raise ValueError(
                f"BCCAEClient base_url must use https:// (got scheme={parsed.scheme!r})"
            )
    else:
        raise ValueError(
            f"BCCAEClient base_url must use https:// (got scheme={parsed.scheme!r})"
        )
    # Reconstruct from parsed components (scheme + netloc only) to break the
    # taint chain from the raw caller-supplied string to the outbound request.
    return urlunparse((parsed.scheme, parsed.netloc, "", "", "", ""))


class BCCAEClient:
    """Synchronous HTTP client for the BCCAE V1 Phase 3 endpoints.

    Each method maps 1:1 to an edge function:
    - :meth:`evaluate`     → v1-bccae-evaluate
    - :meth:`execute`      → v1-bccae-execute
    - :meth:`revoke`       → v1-bccae-revoke
    - :meth:`get_evidence` → v1-bccae-evidence

    Authorization denials are returned (not raised). Network errors,
    invalid API keys, and 5xx responses raise :exc:`~atlasent.AtlaSentError`.

    Use :func:`generate_bccae_nonce` to produce a valid ``caller_nonce``.

    Args:
        api_key: AtlaSent API key with appropriate ``bccae:*`` scopes.
        base_url: Override the API base URL.
        timeout: HTTP request timeout in seconds.
    """

    def __init__(
        self,
        api_key: str,
        *,
        base_url: str = DEFAULT_BASE_URL,
        timeout: float = DEFAULT_TIMEOUT,
    ) -> None:
        if not api_key or not isinstance(api_key, str):
            raise ValueError("BCCAEClient: api_key is required")
        self._api_key = api_key
        self._base_url = _enforce_tls(base_url).rstrip("/")
        self._client = httpx.Client(
            headers={
                "Content-Type": "application/json",
                "Accept": "application/json",
                "Authorization": f"Bearer {api_key}",
                "User-Agent": "atlasent-bccae-python/1.0",
            },
            timeout=timeout,
        )

    def __enter__(self) -> BCCAEClient:
        return self

    def __exit__(self, *_: Any) -> None:
        self.close()

    def close(self) -> None:
        self._client.close()

    # ── Public API ────────────────────────────────────────────────────────────────────

    def evaluate(
        self,
        *,
        actor_id: str,
        actor_type: BccaeActorType,
        actor_trust_level: BccaeTrustLevel,
        action_id: str,
        execution_intent: str,
        caller_nonce: str,
        resource_ref: str,
        resource_type: str,
        resource_classification: BccaeResourceClassification,
        deployment_env: BccaeDeploymentEnv,
        deployment_region: str,
        security_posture: BccaeSecurityPosture,
        actor_claims: dict[str, Any] | None = None,
        organization_version: int | None = None,
        external_signals: list[Any] | None = None,
        dependencies: list[Any] | None = None,
        policy_version_set: list[Any] | None = None,
        request_source: BccaeRequestSource | None = None,
        request_chain_id: str | None = None,
        parent_eval_id: str | None = None,
    ) -> dict[str, Any]:
        """Submit a protected action for BCCAE context construction and evaluation.

        Returns a dict with ``evaluation_id``, ``envelope_hash``,
        ``permit_token``, ``permit_id``, ``expires_at``, and ``outcome``.

        Requires API key with ``bccae:evaluate`` scope.
        """
        body: dict[str, Any] = {
            "actor_id": actor_id,
            "actor_type": actor_type,
            "actor_trust_level": actor_trust_level,
            "action_id": action_id,
            "execution_intent": execution_intent,
            "caller_nonce": caller_nonce,
            "resource_ref": resource_ref,
            "resource_type": resource_type,
            "resource_classification": resource_classification,
            "deployment_env": deployment_env,
            "deployment_region": deployment_region,
            "security_posture": security_posture,
        }
        if actor_claims is not None:
            body["actor_claims"] = actor_claims
        if organization_version is not None:
            body["organization_version"] = organization_version
        if external_signals is not None:
            body["external_signals"] = external_signals
        if dependencies is not None:
            body["dependencies"] = dependencies
        if policy_version_set is not None:
            body["policy_version_set"] = policy_version_set
        if request_source is not None:
            body["request_source"] = request_source
        if request_chain_id is not None:
            body["request_chain_id"] = request_chain_id
        if parent_eval_id is not None:
            body["parent_eval_id"] = parent_eval_id

        return self._post("/v1/bccae/evaluations", body)

    def execute(
        self,
        *,
        permit_token: str,
        action_id: str,
        resource_ref: str,
    ) -> dict[str, Any]:
        """Present a permit token to the Execution Gate (10-check verification).

        Returns a dict with ``authorized`` (bool) and ``outcome``. On denial
        the dict also contains ``check`` and ``reason`` identifying the
        specific gate check that failed.

        Never raises on denial — only raises on network errors or 5xx.
        Requires API key with ``bccae:execute`` scope.
        """
        return self._post(
            "/v1/bccae/execute",
            {
                "permit_token": permit_token,
                "action_id": action_id,
                "resource_ref": resource_ref,
            },
        )

    def revoke(
        self,
        *,
        target_type: BccaeRevocationTargetType,
        target_id: str,
        reason: str,
    ) -> dict[str, Any]:
        """Add an entry to the BCCAE revocation ledger.

        Idempotent — raises :exc:`~atlasent.AtlaSentError` with code
        ``conflict`` if the target is already revoked.

        Requires API key with ``bccae:revoke`` scope.
        """
        return self._post(
            "/v1/bccae/revocations",
            {
                "target_type": target_type,
                "target_id": target_id,
                "reason": reason,
            },
        )

    def get_evidence(self, evidence_id: str) -> dict[str, Any]:
        """Fetch a single evidence record and verify its hash chain integrity.

        ``result["chain_integrity"]["hash_intact"]`` is ``True`` when the
        record's ``record_hash`` matches the server-recomputed value.

        Requires API key with ``bccae:audit`` scope.
        """
        if not evidence_id or not isinstance(evidence_id, str):
            raise ValueError("BCCAEClient: evidence_id is required")
        return self._get(f"/v1/bccae/evidence/{quote(evidence_id, safe='')}")

    # ── HTTP primitives ────────────────────────────────────────────────────

    def _post(self, path: str, body: dict[str, Any]) -> dict[str, Any]:
        try:
            resp = self._client.post(f"{self._base_url}{path}", json=body)
        except httpx.TransportError as exc:
            raise AtlaSentError(
                f"BCCAEClient: network error on POST {path}: {exc}",
                code="network",
            ) from exc
        return self._handle_response(resp, path)

    def _get(self, path: str) -> dict[str, Any]:
        try:
            resp = self._client.get(f"{self._base_url}{path}")
        except httpx.TransportError as exc:
            raise AtlaSentError(
                f"BCCAEClient: network error on GET {path}: {exc}",
                code="network",
            ) from exc
        return self._handle_response(resp, path)

    def _handle_response(self, resp: httpx.Response, path: str) -> dict[str, Any]:
        try:
            data: dict[str, Any] = resp.json()
        except Exception as exc:
            raise AtlaSentError(
                (
                    "BCCAEClient: non-JSON response "
                    f"(status {resp.status_code}) from {path}"
                ),
                code="network",
            ) from exc

        if not resp.is_success:
            message = (
                data.get("message")  # type: ignore[union-attr]
                if isinstance(data, dict)
                else None
            ) or f"BCCAE request failed with status {resp.status_code}"
            code = {
                401: "unauthorized",
                403: "permission_denied",
                404: "not_found",
                409: "conflict",
                429: "rate_limited",
            }.get(resp.status_code, "network")
            raise AtlaSentError(str(message), code=code)  # type: ignore[arg-type]

        return data
